### PR TITLE
[Merged by Bors] - Reduce code duplication by using QueryIterationCursor in QueryIter

### PR DESCRIPTION
**Pull request by james7132**
# Objective
We have duplicated code between `QueryIter` and `QueryIterationCursor`. Reuse that code.

## Solution
 - Reuse `QueryIterationCursor` inside `QueryIter`.
 - Slim down `QueryIter` by removing the `&'w World`. It was only being used by the `size_hint` and `ExactSizeIterator` impls, which can use the QueryState and &Archetypes in the type already.
 - Benchmark to make sure there is no significant regression.

Relevant benchmark results seem to show that there is no tangible difference between the two. Everything seems to be either identical or within a workable margin of error here.

```
group                                          embed-cursor                            main
-----                                          ------------                            ----
fragmented_iter/base                           1.00   387.4±19.70ns        ? ?/sec     1.07   413.1±27.95ns        ? ?/sec
many_maps_iter                                 1.00     27.3±0.22ms        ? ?/sec     1.00     27.4±0.10ms        ? ?/sec
simple_iter/base                               1.00     13.8±0.07µs        ? ?/sec     1.00     13.7±0.17µs        ? ?/sec
simple_iter/sparse                             1.00     61.9±0.37µs        ? ?/sec     1.00     62.2±0.64µs        ? ?/sec
simple_iter/system                             1.00     13.7±0.34µs        ? ?/sec     1.00     13.7±0.10µs        ? ?/sec
sparse_fragmented_iter/base                    1.00     11.0±0.54ns        ? ?/sec     1.03     11.3±0.48ns        ? ?/sec
world_query_iter/50000_entities_sparse         1.08    105.0±2.68µs        ? ?/sec     1.00     97.5±2.18µs        ? ?/sec
world_query_iter/50000_entities_table          1.00     27.3±0.13µs        ? ?/sec     1.00     27.3±0.37µs        ? ?/sec
```

### CODE REVIEW

**Comment by james7132** [comment]
The use of `transmute` with the slice iterator is something I can revert. I just wanted to make the type smaller, but I don't think it's necessary to show identical performance across this change.

File: crates/bevy_ecs/src/query/iter.rs [hunk 1]
```diff
@@ -397,8 +316,11 @@ where
         QueryIterationCursor {
             fetch,
             filter,
-            table_id_iter: query_state.matched_table_ids.iter(),
-            archetype_id_iter: query_state.matched_archetype_ids.iter(),
+            id_iter: if Self::IS_DENSE {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/iter.rs:319]: If we want to keep this, I want solid safety comments explaining what's happening here and why it's safe.

**Review by alice-i-cecile — COMMENTED** [review]
I like the deduplication a lot. The code changes look solid otherwise.

I'm a bit iffy on the transmute, but I'm open to being convinced by good comments.

File: crates/bevy_ecs/src/query/iter.rs (hunk 1 above)
Comment (james7132) [reply crates/bevy_ecs/src/query/iter.rs:321]: Done.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/query/iter.rs [hunk 2]
```diff
@@ -397,8 +316,13 @@ where
         QueryIterationCursor {
             fetch,
             filter,
-            table_id_iter: query_state.matched_table_ids.iter(),
-            archetype_id_iter: query_state.matched_archetype_ids.iter(),
+            // SAFE: The associated IDs are exactly the same size and alignment and
```
Comment (akimakinai) [inline crates/bevy_ecs/src/query/iter.rs:319]: Technically this relies on the unspecified layouts of `TableId`, `ArchetypeId`, and `slice::Iter`, which are all [`repr(Rust)`](https://doc.rust-lang.org/nomicon/transmutes.html) types.
By adding `#[repr(transparent)]` to `TableId` and `ArchetypeId`, it becomes safe to `unsafe{ std::slice::from_raw_parts(_.as_ptr() as *const usize, _.len()) }.iter()` IIUC.

Comment (james7132) [reply crates/bevy_ecs/src/query/iter.rs]: Undoing this because `std::slice::Iter` is  `repr(Rust)`.

**Comment by alice-i-cecile** [comment]
@bevyengine/ecs-team reviews please <3

**Review by TheRawMeatball — APPROVED** [review]

**Pull request closed without merging**